onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -255,6 +259,41 @@ TypeGroup GetTypeGroup(DataType type) {
   return Unknown;
 }
 
+int TypeWidth(DataType type) {
+  if (*type == "tensor(bool)") {
+    return 1;
+  }
+
+  if (*type == "tensor(uint8)" || *type == "tensor(int8)") {
+    return 8;
+  }
+
+  if (*type == "tensor(uint16)" || *type == "tensor(int16)" || *type == "tensor(uint16)" || *type == "tensor(int16)" || *type == "tensor(bfloat16)" || *type == "tensor(float16)") {
+    return 16;
+  }
+
+  if (*type == "tensor(uint32)" || *type == "tensor(int32)" || *type == "tensor(float)") {
+    return 32;
+  }
+
+  if (*type == "tensor(uint64)" || *type == "tensor(int64)" || *type == "tensor(double)") {
+    return 64;
+  }
+
+  return -1;
+}
+
+inline bool LossOfPrecision(DataType src_type, DataType dst_type, const Node& node) {
+  TypeGroup src_type_group = GetTypeGroup(src_type);
+  TypeGroup dst_type_group = GetTypeGroup(dst_type);
+  if (src_type_group == TypeGroup::Unknown || dst_type_group == TypeGroup::Unknown) {
+    return true;
+  }
+  // The comparison with "InsertedPrecisionFreeCast_" reflects cast nodes that are inserted by InsertCastTransforme.
+  // Such casts should not be considered as loss of precision - the inserted upcasts (f16 -> f32) and downcasts (f32 -> f16) are inserted to support kernels when on a CPU EP without F16 support.
+  return (dst_type_group < src_type_group) || (TypeWidth(dst_type) < TypeWidth(src_type) && (node.Name().compare(0, 26, "InsertedPrecisionFreeCast_")));
```

Comment:
This is awkwardly around since the InsertCastTransformer introduces cast nodes with this naming scheme for upcasting inputs to f16 kernels to f32 when using the CPU Execution Provider.